ligo/src/token/fa2.mligo

```diff
@@ -0,0 +1,155 @@
+// SPDX-FileCopyrightText: 2020 TQ Tezos
+// SPDX-License-Identifier: LicenseRef-MIT-TQ
+
+// Corresponds to Token/FA2.hs module
+
+#include "../types.mligo"
+#include "../management.mligo"
+
+
+let unfrozen_token_id: nat = 0n
+let frozen_token_id: nat = 1n
+
+// -----------------------------------------------------------------
+// Helper
+// -----------------------------------------------------------------
+
+[@inline]
+let check_sender (from_ , store : address * storage): address =
+  if (sender = store.admin) then from_
+  else if (sender = from_) then from_
+  else
+    let key: operator = { owner = from_; operator = sender} in
+    if Big_map.mem key store.operators then
+      from_
+    else
+      (failwith("FA2_NOT_OPERATOR"): address)
+
+let debit_from (amt, from_, token_id, store: nat * address * token_id * storage): storage =
+  match Big_map.find_opt (from_, token_id) store.ledger with
+    Some bal ->
+      let updated_store =
+        match Michelson.is_nat (bal - amt) with
+          Some new_bal ->
+            { store with
+              ledger = Big_map.update (from_, token_id) (Some new_bal) store.ledger
+            }
+        | None ->
+            // TODO: How to fail with some value
+            // (failwith ("FA2_INSUFFICIENT_BALANCE", (amt, bal)) : storage)
+            (failwith ("FA2_INSUFFICIENT_BALANCE") : storage)
+      in updated_store
+  | None ->
+      if (amt = 0n) then store // We allow 0 transfer
+      else (failwith ("FA2_INSUFFICIENT_BALANCE") : storage)
+
+let credit_to (amt, to_, token_id, store : nat * address * nat * storage): storage =
+  let new_bal =
+    match Big_map.find_opt (to_, token_id) store.ledger with
+      Some bal -> bal + amt
+    | None -> amt
+  in  { store with
+        ledger = Big_map.update (to_, token_id) (Some new_bal) store.ledger
+      }
+
+
+// -----------------------------------------------------------------
+// Transfer
+// -----------------------------------------------------------------
+
+// TODO: how to add to 'where'
+
+let transfer_item (store, ti : storage * transfer_item): storage =
+  let transfer_one (acc, tx : storage * transfer_destination): storage =
+    let valid_from_ = check_sender (ti.from_, acc) in
+    if tx.amount = 0n then
+        acc
+      else (
+        if (tx.token_id = unfrozen_token_id || (tx.token_id = frozen_token_id && sender = store.admin)) then
+          let updated_store =
+            credit_to
+              ( tx.amount, tx.to_, tx.token_id
+              , debit_from (tx.amount, valid_from_, tx.token_id, acc)
+              )
+          in updated_store
+        else if tx.token_id = frozen_token_id then
+          (failwith ("FROZEN_TOKEN_NOT_TRANSFERABLE"): storage)
+        else
+          (failwith ("FA2_TOKEN_UNDEFINED") : storage)
+      )
+  in
+  let result = List.fold transfer_one ti.txs store
+  in result
```

Comment:
A couple of points here: first the minor ones are that `updated_store` and `result` don't need to be declared and that the top-most `else` is not correctly indented.

More importantly there is an opportunity for optimization: by using the first big condition in the inner `if` you can write its block code once, however this comes with the cost of having to re-check the `tx.token_id = frozen_token_id` later on.
We can refactor this with the use of a `valid_token_id` and save that too, I suggest this (including the minor points above):
```suggestion
    if tx.amount = 0n then acc
    else (
      let valid_token_id =
        if tx.token_id = unfrozen_token_id then tx.token_id
        else if tx.token_id = frozen_token_id then (
          if sender = store.admin then tx.token_id
          else (failwith ("FROZEN_TOKEN_NOT_TRANSFERABLE"): token_id)
        ) else (failwith ("FA2_TOKEN_UNDEFINED") : token_id)
      in credit_to
        ( tx.amount, tx.to_, valid_token_id
        , debit_from (tx.amount, valid_from_, valid_token_id, acc)
        )
    )
  in List.fold transfer_one ti.txs store
```

Reply:
That makes sense, I've made this change as well as other changes as well.